server/worker/jira.go

```diff
@@ -1,31 +1,152 @@
+{{ end }}
+
+View the affected software and more affected hosts:
+
+# Go to the [Software|{{ .FleetURL }}/software/manage] page in Fleet.
+# Above the list of software, in the *Search software* box, enter "{{ .CVE }}".
+# Hover over the affected software and select *View all hosts*.
+
+----
+
+This issue was created automatically by your Fleet Jira integration.
+`,
+))
+
+type jiraTemplateArgs struct {
+	NVDURL   string
+	FleetURL string
+	CVE      string
+	Hosts    []*fleet.HostShort
 }
 
-func NewJira(ds fleet.Datastore, log kitlog.Logger) *Jira {
-	return &Jira{
-		ds:  ds,
-		log: log,
-	}
+// JiraClient defines the method required for the client that makes API calls
+// to Jira.
+type JiraClient interface {
+	CreateIssue(ctx context.Context, issue *jira.Issue) (*jira.Issue, error)
+}
+
+// Jira is the job processor for jira integrations.
+type Jira struct {
+	FleetURL   string
+	Datastore  fleet.Datastore
+	Log        kitlog.Logger
+	JiraClient JiraClient
 }
 
+// Name returns the name of the job.
 func (j *Jira) Name() string {
-	return "jira"
+	return JiraName
+}
+
+// JiraArgs are the arguments for the Jira integration job.
+type JiraArgs struct {
+	CVE string `json:"cve"`
 }
 
+// Run executes the jira job.
 func (j *Jira) Run(ctx context.Context, argsJSON json.RawMessage) error {
-	// TODO: implement me
+	var args JiraArgs
+	if err := json.Unmarshal(argsJSON, &args); err != nil {
+		return ctxerr.Wrap(ctx, err, "unmarshal args")
+	}
+
+	hosts, err := j.Datastore.HostsByCVE(ctx, args.CVE)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "find hosts by cve")
+	}
+
+	tmplArgs := jiraTemplateArgs{
+		NVDURL:   nvdCVEURL,
+		FleetURL: j.FleetURL,
+		CVE:      args.CVE,
+		Hosts:    hosts,
+	}
+
+	var buf bytes.Buffer
+	if err := jiraSummaryTmpl.Execute(&buf, &tmplArgs); err != nil {
+		return ctxerr.Wrap(ctx, err, "execute summary template")
+	}
+	summary := buf.String()
+
+	buf.Reset() // reuse buffer
+	if err := jiraDescriptionTmpl.Execute(&buf, &tmplArgs); err != nil {
+		return ctxerr.Wrap(ctx, err, "execute description template")
+	}
+	description := buf.String()
+
+	// Note, newlines get automatically escaped in json.
+
+	issue := &jira.Issue{
+		Fields: &jira.IssueFields{
+			Type: jira.IssueType{
+				Name: "Bug",
+			},
+			Summary:     summary,
+			Description: description,
+		},
+	}
+
+	createdIssue, err := j.JiraClient.CreateIssue(ctx, issue)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "create issue")
+	}
+
+	level.Debug(j.Log).Log(
+		"msg", "created jira issue for cve",
+		"cve", args.CVE,
+		"issue_id", createdIssue.ID,
+		"issue_key", createdIssue.Key,
+	)
+
+	return nil
+}
+
+// QueueJiraJobs queues the Jira vulnerability jobs to process asynchronously
+// via the worker.
+func QueueJiraJobs(ctx context.Context, ds fleet.Datastore, logger kitlog.Logger, recentVulns map[string][]string) error {
+	level.Debug(logger).Log("enabled", "true", "recentVulns", len(recentVulns))
```

Comment:
For troubleshooting purposes: I would turn this line into `INFO`, and add a separate `DEBUG` that prints the full list of recentVulns (in one line).

Reply:
That's potentially a lot of info, I don't think CPEs are particularly useful, how about I log only the CVEs in debug mode?

Reply:
+1 to CVEs, just something you think it would be helpful in case a user hits an issue and they turn `--logging_debug` on to troubleshoot.